cosmo_tester/test_suites/test_manager_upgrade/manager_upgrade_test.py

```diff
@@ -0,0 +1,379 @@
+BOOTSTRAP_BRANCH = '3.4m5'
+
+UPGRADE_REPO_URL = 'https://github.com/cloudify-cosmo/'\
+                   'cloudify-manager-blueprints.git'
+UPGRADE_BRANCH = 'master'
+
+
+class ManagerUpgradeTest(TestCase):
+
+    def test_manager_upgrade(self):
+        """Bootstrap a manager, upgrade it, rollback it, examine the results.
+
+        To test the manager in-place upgrade procedure:
+            - bootstrap a manager (this is part of the system under test,
+              does destructive changes to the manager, and need a known manager
+              version: so, can't use the testenv manager)
+            - deploy the hello world app
+            - upgrade the manager, changing some inputs (eg. the port that
+              Elasticsearch uses)
+            - check that everything still works (the previous deployment still
+              reports metrics; we can install another deployment)
+            - rollback the manager
+            - post-rollback checks: the changed inputs are now the original
+              values again, the installed app still reports metrics
+        """
+        self.prepare_manager()
+
+        self.preupgrade_deployment_id = self.deploy_hello_world('pre-')
+
+        self.upgrade_manager()
+        self.post_upgrade_checks()
+
+        self.rollback_manager()
+        self.post_rollback_checks()
+
+        self.teardown_manager()
+
+    @contextmanager
+    def _manager_fabric_env(self):
+        inputs = self.manager_inputs
+        with fabric.context_managers.settings(
+                host_string=self.upgrade_manager_ip,
+                user=inputs['ssh_user'],
+                key_filename=inputs['ssh_key_filename']):
+            yield fabric.api
+
+    def _bootstrap_local_env(self, workdir):
+        storage = local.FileStorage(
+            os.path.join(workdir, '.cloudify', 'bootstrap'))
+        return local.load_env('manager', storage=storage)
+
+    def _blueprint_rpm_versions(self, workdir):
+        """RPM filenames that should be installed on the manager.
+        """
+        env = self._bootstrap_local_env(workdir)
+        storage = env.storage
+
+        amqp_influx_rpm = storage.get_node('amqp_influx')['properties'][
+            'amqpinflux_rpm_source_url']
+        restservice_rpm = storage.get_node('rest_service')['properties'][
+            'rest_service_rpm_source_url']
+        mgmtworker_rpm = storage.get_node('mgmt_worker')['properties'][
+            'management_worker_rpm_source_url']
+        return {
+            'cloudify-amqp-influx': amqp_influx_rpm,
+            'cloudify-rest-service': restservice_rpm,
+            'cloudify-management-worker': mgmtworker_rpm
+        }
+
+    def _cloudify_rpm_versions(self):
+        with self._manager_fabric_env() as fabric:
+            return fabric.sudo('rpm -qa | grep cloudify')
+
+    def check_rpm_versions(self):
+        blueprint_rpms = self._blueprint_rpm_versions(self.cfy_workdir)
```

Comment:
No, just avoiding coupling the directory to the `_blueprint_rpm_versions` function - that function is fairly generic and doesn't assume things..